### ISSUE TITLE
Keep deskband after restart

### ISSUE THREAD

**Issue by dsafa** [dsafa/CSDeskBand#25]
Not sure if this can be done or not but something to look into.

**Comment by dsafa** [comment]
This is done by other deskbands but I am not sure how or what is missing

**Comment by dsafa** [comment]
Not sure if its just a bug in the current windows 10 version but toolbars only persist sometimes when resizing or moving a deskband. Even for the built in toolbars.

**Comment by 20154530** [comment]
Did you find out what caused it,I also find that it must be moved to leftside and then move back (horizontal mode),after that it will possibly remains open when restart

**Comment by dsafa** [comment]
Never found out the cause, but in my experience now, it works properly if you just leave it for a bit. I know windows writes the current toolbars to the registry so maybe theres some delays or it does it every set interval.

**Comment by 20154530** [comment]
I find another DeskBand Project on Github and its Sample works fun : [sharpshell
](https://github.com/dwmkerr/sharpshell#desk-band-extensions).I compared your codes and that one(sharpshell) ,and find that some of your [COMImprot] functions do not have return as that does,so I copied his [COMImport] interfaces,and now it can keep alive when restart.

**Comment by dsafa** [comment]
Your English is fine. Thanks for looking into it. Do you have a link to which COM interfaces are different?

**Comment by dsafa** [comment]
I updated the interfaces and it works fine for me. I checked the registry `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\Streams\Desktop` after selecting a toolbar and it appears to be saved properly

**Comment by xmoer** [comment]
I spent a lot of time exploring the root causes of this problem. It's worth celebrating that I found a solution. First of all it is caused by the BUG of the system itself. Secondly, the problem lies in the language status bar. Third, the toolbar of the system itself has similar problems.

Windows7's language status bar is on the left, where deskband is currently located. But the initial restart state of Windows10 is actually the same as that of Windows7. The language status bar still defaults to the left and then automatically moves to the right. This process creates the problem of blocking the toolbar.

In fact, toolbars still exist, deskband is working, but they are invisible. So we just need to force the task bar to refresh at the right time. For example:
```
SendNotifyMessage((IntPtr)HWND_BROADCAST, (uint)WindowsMessage.WM_SETTINGCHANGE, UIntPtr.Zero, "TraySettings");
```
Personally, I suggest that the above code can be sent several times after the system startup. The problem of keeping deskband after restart has been solved, and the project should be excellent now.

**Comment by xmoer** [comment]
Since there are many self-startup programs when the operating system is booted, it is not recommended that the author build the above solution in "CSDeskBand" under such complicated circumstances.

**Comment by dsafa** [comment]
Interesting. Thanks for the details. Let me pin this issue for visibility.

**Comment by MattBlessinger** [comment]
I currently have this problem. The deskband does indeed start, it's just invisible. I tried the SendNotifyMessage method (confirmed a true boolean was returned), but the deskband stayed invisible. Right now the only thing working is to restart explorer through task manager and then it's visible.

**Comment by xmoer** [comment]
> I currently have this problem. The deskband does indeed start, it's just invisible. I tried the SendNotifyMessage method (confirmed a true boolean was returned), but the deskband stayed invisible. Right now the only thing working is to restart explorer through task manager and then it's visible.

I have tested several versions of Windows10, and all of the above methods are effective. You should send a "SendNotifyMessage" when the toolbar is not visible, but we don't know when the toolbar will disappear. So you can use "**Timer()**" to send messages continuously within five minutes of the Windows10 startup. It must be noted that sending only once is absolutely not enough.

**Comment by MattBlessinger** [comment]
> I have tested several versions of Windows10, and all of the above methods are effective. You should send a "SendNotifyMessage" when the toolbar is not visible, but we don't know when the toolbar will disappear. So you can use "**Timer()**" to send messages continuously within five minutes of the Windows10 startup. It must be noted that sending only once is absolutely not enough.

I made a standalone application that I ran at startup. Every 2 seconds it sends the "SendNotifyMessage" command. I let that run for a couple of minutes, but the deskband still did not appear. Can you share your code so that I may test it?

**Comment by xmoer** [comment]
> I made a standalone application that I ran at startup. Every 2 seconds it sends the "SendNotifyMessage" command. I let that run for a couple of minutes, but the deskband still did not appear. Can you share your code so that I may test it?

I'm sure you didn't quote my code, but wrote your own global "BROADCAST", right? The key to success is not the number of messages sent, but the parameters I set. Your "SendNotifyMessage" code must be the same as mine. This is a message code that is only valid for Windows10, and Windows10 has a large number of Undocumented APIs.

**Comment by xmoer** [comment]
> I made a standalone application that I ran at startup. Every 2 seconds it sends the "SendNotifyMessage" command. I let that run for a couple of minutes, but the deskband still did not appear. Can you share your code so that I may test it?

I just wrote a simple "Timer()" and sent messages every 15 seconds, a total of 12 times. There is no doubt that the problem is not a timer. I tested almost all Windows10 versions and they worked perfectly.

**Comment by MattBlessinger** [comment]
> I'm sure you didn't quote my code, but wrote your own global "BROADCAST", right? The key to success is not the number of messages sent, but the parameters I set. Your "SendNotifyMessage" code must be the same as mine. This is a message code that is only valid for Windows10, and Windows10 has a large number of Undocumented APIs.

This is what I did. What values did you use for HWND_BROADCAST and WM_SETTINGCHANGE?
```
private const int HWND_BROADCAST = 0xffff;
private const int WM_SETTINGCHANGE = 0x001a;
bool retVal = SendNotifyMessage((IntPtr)HWND_BROADCAST, WM_SETTINGCHANGE, UIntPtr.Zero, "TraySettings");
```

**Comment by xmoer** [comment]
> This is what I did. What values did you use for HWND_BROADCAST and WM_SETTINGCHANGE?

Yeah, it seems all right. The most important "TraySettings" parameter settings are fine. It seems that we need to discuss our respective development environments. 

I used: CSDeskBand Version 3.1 + WPF + VS2017.
If you don't mind, could you provide me with your file to help you find out the cause of the failure?

**Comment by MattBlessinger** [comment]
> I used: CSDeskBand Version 3.1 + WPF + VS2017.
> If you don't mind, could you provide me with your file to help you find out the cause of the failure?

I used CSDeskBand v3.1, WPF .NET Framework 4.7.2, VS2019. I attached the separate program I'm using to send the broadcast. I run the .exe in /bin/debug at startup. It sends the broadcast every 5 seconds until you press a key in the console window. Thank you for your help!
[ConsoleApp2.zip](https://github.com/dsafa/CSDeskBand/files/3667391/ConsoleApp2.zip)

**Comment by xmoer** [comment]
> I used CSDeskBand v3.1, WPF .NET Framework 4.7.2, VS2019. I attached the separate program I'm using to send the broadcast. I run the .exe in /bin/debug at startup. It sends the broadcast every 5 seconds until you press a key in the console window. Thank you for your help!

It's strange that it works well on my computer.

**Comment by MattBlessinger** [comment]
> It's strange that it works well on my computer.

Well that's unfortunate for me then. The deskband stays invisible unless I restart explorer.exe

**Comment by xmoer** [comment]
> Well that's unfortunate for me then. The deskband stays invisible unless I deselect and reselect it from the menu or restart explorer.exe

Restarting the desktop process is effective, which shows that my point of view is correct. I even tested Windows Server 2016 on a virtual machine, and your program is still worked well. So maybe the error originated from your DLL file?

**Comment by MattBlessinger** [comment]
> So maybe the error originated from your DLL file?

Potentially. Did you add anything extra to your deskband in terms of window visiblity, etc., that could affect it? I attached my deskband so you may try it on your computer.

**Comment by xmoer** [comment]
> Potentially. Did you add anything extra to your deskband in terms of window visiblity, etc., that could affect it? I attached my deskband so you may try it on your computer.

This DLL file may have problems. There is no "TaskbarInfo" event in your code, and I guess that's why it can't receive messages. The DLL file can't even work after Windows Server 2016 system restart. I think you should study the author's demo carefully.

```
public class Deskband : CSDeskBandWpf
{
    ……
}
```

Don't let the content here be empty. The author of "CSDeskBand" has another excellent project that you can refer to: https://github.com/dsafa/audio-band

**Comment by MattBlessinger** [comment]
I found the issue after adding my code piece by piece to the ExampleWpf until it stopped working after reboot. It turns out it was this line I was using in UserControl to programmatically set the height to be the taskbar height.
`this.Height = System.Windows.SystemParameters.PrimaryScreenHeight - System.Windows.SystemParameters.WorkArea.Height;`
FYI, I was using this instead of TaskbarInfo.Size.Height, because TaskbarInfo does not account for resolution scaling. 

Since the deskband is just for myself I have hard coded the height of the UserControl, and now the deskband is visible after rebooting! The reason it wasn't working after reboot is that the taskbar did not exist by the time the deskband started so it set the UserControl height to 0.

To confirm it was user error and there are no problems with CSDeskBand. Thank you for the help.

**Comment by kevinglier** [comment]
@MattBlessinger I think it could help, overriding the event callbacks OnInitialized and OnRenderSizeChanged, like I mentioned in https://github.com/dsafa/CSDeskBand/issues/51.

Thanks for you comment by the way, which was helpful for me. :)

**Comment by eladpotok** [comment]
So there is any solution?
My issue is not that the toolbar is now shown after restart but is not selected.
When I open the toolbar selection I don't see the checkmark (v) nearby.
